src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -332,12 +332,36 @@ private int commonPrefixLength(List<String> classNames) {
     for (int i = 0; i < size; i++) {
       String a = classNames.get(i);
       String b = typeSpecStack.get(i).name;
-      if (!a.equals(b)) return i;
+      if (!a.equals(b)) {
+        for (; i > 0; i--) {
+          // Does the current type have a parent type with a member type with the same name
+          if (!parentTypeHasMemberType(classNames.get(i), i)) {
```

Comment:
Looks wrong. My guess is you want to invert this: if any parent has a member type return 0; if none do return i.